### PR TITLE
[BEAM-11105] Add docs + CHANGES.md entry for Go Watermark Estimation

### PR DESCRIPTION
**Pull request by damccorm**
Now that all work required for watermark estimation has been added, we can update the docs + changelog to let people know how to use it!

Completed prs - #17267, #17374, #17475, #17476

Associated design doc - https://docs.google.com/document/d/1DqCYJ-J1YGNelCRIcN5v6BQsZxJB2l5uWBmN4ti--Ew/edit?usp=sharing

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by damccorm** [comment]
R: @jrmccluskey since Ritesh is taking the day off

**Review by jrmccluskey — APPROVED** [review]
LGTM!

**Comment by damccorm** [comment]
R: @lostluck

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Review by lostluck — CHANGES_REQUESTED** [review]

File: sdks/go/examples/snippets/04transforms.go [hunk 1]
```diff
@@ -105,6 +106,52 @@ func (fn *splittableDoFn) ProcessElement(element string, bf beam.BundleFinalizat
 
 // [END bundlefinalization_simplecallback]
 
+// [START watermarkestimation_customestimator]
+
+// (Optional) - define a custom watermark state type.
```
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/examples/snippets/04transforms.go:111]: Eg. Especially where we have "default" state types and estimators, this is a good place to list them...

[the remark this answers is not in the document]
Comment (damccorm) [reply sdks/go/examples/snippets/04transforms.go]: That's fine - we aren't particularly consistent on that with our docs though. Might be worth a pass to do that at some point. I updated to do this.

**Review by damccorm — COMMENTED** [review]
Thanks for the feedback - updated accordingly!

File: sdks/go/examples/snippets/04transforms.go [hunk 2]
```diff
@@ -105,6 +108,64 @@ func (fn *splittableDoFn) ProcessElement(element string, bf beam.BundleFinalizat
...
+	return WatermarkState{Watermark: time.Now()}
+}
+
+// CreateWatermarkEstimator creates the watermark estimator used by this Splittable DoFn.
+// Must take in a state parameter if InitialWatermarkEstimatorState is defined, otherwise takes no parameters.
+func (fn *weDoFn) CreateWatermarkEstimator(initialState WatermarkState) *CustomWatermarkEstimator {
+	return &CustomWatermarkEstimator{state: initialState}
+}
+
+// WatermarkEstimatorState returns the state usedto resume future watermark estimation
```
Comment (lostluck) [inline sdks/go/examples/snippets/04transforms.go:153]:
```suggestion
// WatermarkEstimatorState returns the state used to resume future watermark estimation
```

**Review by lostluck — APPROVED** [review]
I spotted a typo, but I'll apply that fix, then merge.
Thanks and congratulations!

**Pull request merged**